src/main/scala/gen/gmos/GmosNLongslitD.scala

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2019 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package basic
+package gen.gmos
+
+import basic.search.Constraints
+import basic.syntax.all._
+
+import gem.Step
+import gem.config.DynamicConfig.GmosN
+import gem.config.GmosConfig._
+import gem.enum._
+import gem.math.{ MagnitudeValue, Wavelength }
+
+import cats.Functor
+import cats.effect.Sync
+import cats.implicits._
+import fs2.Stream
+import monocle.{Prism, Optional}
+
+import java.time.Duration
+import java.util.concurrent.TimeUnit.NANOSECONDS
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Dynamic sequence generation for GMOS North Longslit.
+ */
+sealed trait GmosNLongslitD {
+
+  /**
+   * Generates a sequence that is as long as necessary to achieve the desired
+   * s/n ratio.  Integration time is adjusted throughout to match conditions.
+   *
+   * @param itc integration time calculator
+   * @param acquired a program that can be evaluated to determine whether an
+   *                 acquisition (or reacquisition) is necessary
+   * @param reachedS2N a program that can be evaluated to determine whether the
+   *                   desired s/n ratio has been reached
+   * @tparam F effect
+   *
+   * @return a sequence that acquires (and reacquires as necessary) the target
+   *         and collects science data and calibrations until the desired s/n
+   *         ratio is achieved
+   */
+  def sequence[F[_]: Sync](
+    itc:        Itc[F],
+    acquired:   F[Boolean],
+    reachedS2N: F[Boolean]
+  ): Stream[F, Step.GmosN]
+
+}
+
+
+object GmosNLongslitD {
+
+  object exposureTime {
+
+    // Maximum exposure time we can handle
+    val Max: Duration =
+      Duration.ofNanos(Long.MaxValue)
+
+    private def toFiniteDuration(d: Duration): Option[FiniteDuration] =
+      Some(d).filter(_.compareTo(Max) <= 0).as(FiniteDuration(d.toMillis, NANOSECONDS))
+
+    private def toDuration(fd: FiniteDuration): Duration =
+      Duration.ofNanos(fd.toNanos)
+
+    // Converter between Java and Scala durations.
+    val duration: Prism[Duration, FiniteDuration] =
+      Prism(toFiniteDuration)(toDuration)
+
+    // Optional to set exposure time in a step.
+    val optional: Optional[Step.GmosN, FiniteDuration] =
+      Step.GmosN.dynamicConfig composeLens GmosN.exposureTime composePrism duration
+
+    // Lookup the exposure time and set it in the given step.
+    def evalSet[F[_]: Sync](time: F[FiniteDuration])(s: Step.GmosN): F[Step.GmosN] =
+      time.map { optional.set(_)(s) }
+
+  }
+
+  // TODO: somewhere else?
+  final implicit class StreamOps[F[_], O](s: Stream[F, O]) {
```

Comment:
Contribute back to fs2 maybe?

Reply:
I suppose I could ask on the FS2 channel to see if they would be interested in something like that.  Maybe there's a better way and this is dumb since I'm new to this.  It seemed pretty cool when it occurred to me though.

Reply:
Yeah you should ask. They're nice.